### PR TITLE
2.2.6 Use batches and direct queries to fix sales address upgrade

### PR DESCRIPTION
**Pull request by rikwillems**
### Description (*)
Magento Commerce compatible fix for sales address upgrade script.

### Fixed Issues (if relevant)
1. magento/magento2#16570: enhance performance on large catalog
[(comment)](https://github.com/magento/magento2/pull/16570#issuecomment-429830198)
2. https://github.com/magento/magento2/pull/18945

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @rikwillems. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.2-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/Sales/Setup/UpgradeData.php [hunk 1]
```diff
@@ -178,27 +179,77 @@ private function convertSerializedDataToJson($setupVersion, SalesSetup $salesSet
...
+     * @param ModuleDataSetupInterface $setup
+     * @param string $address_type
+     */
+    public function fillQuoteAddressIdInSalesOrderAddressByType(ModuleDataSetupInterface $setup, $address_type)
+    {
+        $salesConnection = $setup->getConnection('sales');
+        $orderTable = $setup->getTable('sales_order', 'sales');
+        $orderAddressTable = $setup->getTable('sales_order_address', 'sales');
+
+        $query = $salesConnection
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Setup/UpgradeData.php:181]: Checked, this part generates following SQL example:
```sql
SELECT sales_order_address.entity_id, sales_order_address.address_type, sales_order.quote_id as quote_id
FROM sales_order_address sales_order_address
INNER JOIN sales_order sales_order ON (sales_order_address.parent_id = sales_order.entity_id)
WHERE sales_order_address.quote_address_id IS NULL
  AND sales_order_address.address_type = 'shipping'
ORDER BY sales_order_address.entity_id;
```

Result example:
```php
[
    'order address id' => [
        'entity_id' => 'order address id',
        'address_type' => 'order address type',
        'quote_id' => 'quote id'
    ],
    ...
]

File: app/code/Magento/Sales/Setup/UpgradeData.php [hunk 2]
```diff
@@ -178,27 +179,77 @@ private function convertSerializedDataToJson($setupVersion, SalesSetup $salesSet
...
+     * @param ModuleDataSetupInterface $setup
+     * @param array $orderAddresses
+     * @param string $address_type
+     */
+    public function fillQuoteAddressIdInSalesOrderAddressProcessBatch(ModuleDataSetupInterface $setup, array $orderAddresses, $address_type)
+    {
+        $salesConnection = $setup->getConnection('sales');
+        $quoteConnection = $setup->getConnection('checkout');
+
+        $query = $quoteConnection
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Setup/UpgradeData.php:226]: This query will be following 
```sql
SELECT quote_address.quote_id, quote_address.address_id
FROM quote_address quote_address
INNER JOIN quote quote ON (quote_address.quote_id = quote.entity_id)
WHERE quote.entity_id in (<order_address_ids>)
  AND address_type = 'shipping'
```


Looks like you did some mistake in "<order_address_ids>". Please check it

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
Also looks like CLA showing as "not signed", so probably we have some issue there. Could you contact me in Slack to get it resolved faster?

**Comment by orlangur** [comment]
@ihor-sviziev is this not relevant or already delivered to 2.3?

**Comment by ihor-sviziev** [comment]
@orlangur this PR was just re-created from https://github.com/magento/magento2/pull/18945 by author.

BTW we have the same issue in 2.3:
https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Sales/Setup/Patch/Data/FillQuoteAddressIdInSalesOrderAddress.php

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=19098) <br/>All committers have signed the CLA.

**Comment by rikwillems** [comment]
This CLA is giving me a headache. I tried a new PR fix this as I cannot modify existing commits with new emailaddress. https://github.com/magento/magento2/pull/19114

**Comment by orlangur** [comment]
@rikwillems,

> cannot modify existing commits with new email address

```
git reset --soft HEAD^
git commit -m'Message'
git force push
```

should do the trick for arbitrary amount of last commits.

File: app/code/Magento/Sales/Setup/UpgradeData.php [hunk 3]
```diff
@@ -173,32 +173,91 @@ private function convertSerializedDataToJson($setupVersion, SalesSetup $salesSet
...
     /**
      * Fill quote_address_id in table sales_order_address if it is empty.
-     *
      * @param ModuleDataSetupInterface $setup
      */
     public function fillQuoteAddressIdInSalesOrderAddress(ModuleDataSetupInterface $setup)
     {
-        $addressTable = $setup->getTable('sales_order_address');
-        $updateOrderAddress = $setup->getConnection()
+        $this->fillQuoteAddressIdInSalesOrderAddressByType($setup, Address::TYPE_SHIPPING);
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Setup/UpgradeData.php:166]: In integration tests we have following fail:
> PHP Fatal error:  Uncaught Error: Class 'Magento\Sales\Setup\Address' not found in 

Please import `Address` class

File: app/code/Magento/Sales/Setup/UpgradeData.php [hunk 4]
```diff
@@ -173,32 +173,91 @@ private function convertSerializedDataToJson($setupVersion, SalesSetup $salesSet
...
+
+            $this->fillQuoteAddressIdInSalesOrderAddressProcessBatch($setup, $result, $addressType);
+        }
+    }
+    /**
+     * @param ModuleDataSetupInterface $setup
+     * @param array $orderAddresses
+     * @param string $addressType
+     */
+    public function fillQuoteAddressIdInSalesOrderAddressProcessBatch(ModuleDataSetupInterface $setup, array $orderAddresses, $addressType)
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Setup/UpgradeData.php]: According to static tests failure - this line is too long. Please format this function declaration in few lines

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
Hi @rikwillems,
Could you update your PR by squashing all commits into single one and force push, as @orlangur  showed above?

Also please create PR to 2.3-develop branch, as we should accept it at first according to [contributor guide](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#rules)

**Comment by rikwillems** [comment]
@ihor-sviziev I tried this now. Please see it that works.

**Comment by ihor-sviziev** [comment]
Hi @rikwillems,
Looks like you pushed without force option, as result - it merged changes.
Use `git push origin <your branch> -f`

**Comment by rikwillems** [comment]
@ihor-sviziev this is quote a pain. Not sure why it is not working out. Tried it again.

**Comment by ihor-sviziev** [comment]
@rikwillems, now you force-pushed, but pushed the same commits as before, but we want to have one squashed commit from you.
In order to verify if you did everything is ok - on the commits section all commit authors should be clickable, now we have two commits from email not linked to your github account
https://github.com/magento/magento2/pull/19098/commits

**Comment by rikwillems** [comment]
@ihor-sviziev this should do it...

File: app/code/Magento/Sales/Setup/UpgradeData.php [hunk 5]
```diff
@@ -66,26 +49,17 @@ class UpgradeData implements UpgradeDataInterface
...
-     * @param AddressCollectionFactory $addressCollFactory
-     * @param OrderFactory $orderFactory
-     * @param QuoteFactory $quoteFactory
      * @param State $state
      */
     public function __construct(
         SalesSetupFactory $salesSetupFactory,
         Config $eavConfig,
         AggregatedFieldDataConverter $aggregatedFieldConverter,
-        AddressCollectionFactory $addressCollFactory,
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Setup/UpgradeData.php]: Removing constructor dependencies is not backward compatible change, please revert it

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Issues with commit author is gone, that's good 👍
One small thing still there. Please check it.
Also please create PR to 2.3-develop branch, because it should be merged at first

**Comment by ihor-sviziev** [comment]
Hi @rikwillems,
Are you interested in finalizing this PR?

**Comment by ihor-sviziev** [comment]
Created PR https://github.com/magento/magento2/pull/20785 with the same changes to 2.3-develop branch, as it should be merged first.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-4344](https://jira.corp.magento.com/browse/ENGCOM-4344) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
@rikwillems thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by p-bystritsky** [comment]
@rikwillems @ihor-sviziev there is not such column as "orderAddressId" in "sales_order_address" table, used in 'where' clause in https://github.com/magento/magento2/pull/19098/files#diff-0815805b2d40fcf0d22ebd639e0d73c4R247. Which will lead to an error during update.

File: app/code/Magento/Sales/Setup/UpgradeData.php [hunk 6]
```diff
@@ -173,32 +159,95 @@ private function convertSerializedDataToJson($setupVersion, SalesSetup $salesSet
...
+            ->where('address_type = ?', $addressType);
+
+        $quoteAddresses = $quoteConnection->fetchAssoc($query);
+
+        foreach ($orderAddresses as $orderAddress) {
+            $bind = [
+                'quote_address_id' => $quoteAddresses[$orderAddress['quote_id']]['address_id'] ?? null,
+            ];
+            $where = [
+                'orderAddressId' => $orderAddress['entity_id']
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Setup/UpgradeData.php]:
```suggestion
                'entity_id' => $orderAddress['entity_id']
```

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-4344](https://jira.corp.magento.com/browse/ENGCOM-4344) has been created to process this Pull Request

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-4344](https://jira.corp.magento.com/browse/ENGCOM-4344) has been created to process this Pull Request

**Pull request merged**